EnumTypeValue, EnumValue

maxqchen opened the pull request:
Similar to Tyson's comment, I wasn't sure what the "Unspecified" stuff was all about. The implementation currently should let us do everything specified in https://github.com/whdl/whdl/wiki/Language-Specification#wiki-enums with the exception of explicit conversions from Value -> EnumValue, ie. TrafficLightDelay delay3 = 5 which I don't think should be a thing.

I think I'm still missing a little bit of testing with the EnumValue's equal() method.

maxqchen committed:
EnumTypeValue, EnumValue

lucaswoj commented:
> I wasn't sure what the "Unspecified" stuff was all about

This is a useful feature in the long run, but we don't need to address that in this pull request.

> with the exception of explicit conversions from Value -> EnumValue, ie. TrafficLightDelay delay3 = 5 which I don't think should be a thing.

We agreed on this feature in our language design meetings. If you want to reopen the issue, I'm glad to discuss it at our next team meeting. 

> I think I'm still missing a little bit of testing with the EnumValue's equal() method.

Wonder no more! Just run `gradle jacoco` and open up the generated report :smile:

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/EnumTypeValue.java:
```diff
@@ -0,0 +1,48 @@
+
+public class EnumTypeValue extends TypeValue {
+	
+	private TypeValue typeOfEnums;
```
The name `enumsType` is slightly shorter and cleaner

mtrberzi commented:
This looks good to me. Should be an easy merge once the suggested fixes have been made.

maxqchen commented:
So in terms of a generic values being implicitly converted into an enum... what is the use case for it? The purpose of enum is to hide the actual underlying values in a set and we are already somewhat breaking that by allowing explicit mappings to arbitrary value types.Also, I feel like allowing this assignment will open up a whole new can of worms. For example an enum can then be assigned to a value that only gets evaluated at runtime?

maxqchen committed:
Renamed typeOfEnums -> enumsType
Added equality test for enums

maxqchen commented:
Okay, I updated the diff to address the 2 comments. I'm going to hold off on the arbitrary value -> enum conversion since I didn't hear back from it yet and we should probably discuss it in a meeting. That can be added at a later time.

lucaswoj commented:
@maxqchen Have you run this through jacoco yet? (Forget the exact commands off the top of my head -- `gradle jacoco` and then derp around in `reports/` until you find the jacoco report) Does it have full test coverage?

lucaswoj commented:
I'm ok with leaving off the enum conversion stuff for now. We need a general solution for casting values (probably just some kind of `cast(TypeType type)` method on `Value`) before that's possible anyway.

maxqchen commented:
Is this s‎omething I need to install? I must have missed it. I'll download and run it when I get home.

lucaswoj commented:
`gradle` will handle all the dependencies for you :smile:

maxqchen committed:
Added tests for 100% coverage

maxqchen commented:
BTW this is at 100% coverage now.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/EnumTypeValue.java:
```diff
@@ -0,0 +1,48 @@
+	private Map<String, Value> enums;
+	
+	public EnumTypeValue(TypeValue typeOfEnums, Map<String, Value> enums) throws TypeMismatchException {
+		this.enumsType = typeOfEnums;
```
Rename the parameter to `enumsType` as well

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/EnumTypeValue.java:
```diff
@@ -0,0 +1,48 @@
+		
+		for(Map.Entry<String, Value> e : enums.entrySet()) {
+			this.enums.put(e.getKey(), e.getValue());
+		}
```
Is there a reason you don't just make a shallow copy of the original `enums`?

```
this.enums = new HashMap<String, Value>(enums);
```

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/EnumTypeValue.java:
```diff
@@ -0,0 +1,48 @@
+		return enums.get(key);
+	}
+	
+	public TypeValue getTypeOfEnums() {
```
`getEnumsType`

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/EnumTypeValue.java:
```diff
@@ -0,0 +1,48 @@
+	public void verify() throws Exception{
+		enumsType.verify();
+		for(Value v : enums.values()) {
+			if(!enumsType.equals(v.getType())) {
```
This is good for now. In the near future, I really want to create some explicitly-subclass-aware type comparison methods.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/NoSuchEnumIdentifierException.java:
```diff
@@ -0,0 +1,19 @@
+package org.whdl.frontend.syntaxtree;
+
+public class NoSuchEnumIdentifierException extends Exception {
```
I think renaming this class `EnumIdentifierNotDefined` would be more consistent with our other exception class names

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/EnumTypeValue.java:
```diff
@@ -0,0 +1,48 @@
+	
+	public Value get(String key) {
+		return enums.get(key);
+	}
```
I think that `contains` and `get` are a reinventing the wheel. Why not expose the `this.enums` object directly via a `getEnums` method?

lucaswoj replied:
cc @maxqchen I would like if we just dropped `contains` and `get` methods entirely, exposing a `Map<String, Value> getValues()` method instead.

maxqchen replied:
I didn't want to expose the whole hashmap since then we have to worry about making it immutable or making a copy and what not. it also exposes too much of the object which I don't think is necessary. Someone using the class shouldn't have to care what the underlying representation is nor do they want it. Just my 2c.

lucaswoj replied:
The only entity "using" this class is the compiler which needs to know all the keys and values supported :-) 

> On Feb 19, 2014, at 9:03 PM, Max Qitian Chen notifications@github.com wrote:
> 
> In src/main/java/org/whdl/frontend/syntaxtree/EnumTypeValue.java:
> 
> > -   public EnumTypeValue(TypeValue typeOfEnums, Map<String, Value> enums) throws TypeMismatchException {
> > -       this.enumsType = typeOfEnums;
> > -       this.enums = new HashMap<String, Value>();
> > -   
> > -       for(Map.Entry<String, Value> e : enums.entrySet()) {
> > -           this.enums.put(e.getKey(), e.getValue());
> > -       }
> > -   }
> > -   
> > -   public boolean contains(String key) {
> > -       return enums.containsKey(key);
> > -   }
> > -   
> > -   public Value get(String key) {
> > -       return enums.get(key);
> > -   }
> >   I didn't want to expose the whole hashmap since then we have to worry about making it immutable or making a copy and what not. it also exposes too much of the object which I don't think is necessary. Someone using the class shouldn't have to care what the underlying representation is nor do they want it. Just my 2c.
> 
> —
> Reply to this email directly or view it on GitHub.

maxqchen replied:
Still, knowing who consumes a class doesn't mean we should expose more than we need to. I don't see why we need to provide the entire map. Can you give an example of where the whole map is required and the provided interface isn't sufficient?

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/EnumTypeValue.java:
```diff
@@ -0,0 +1,48 @@
+		return enumsType;
+	}
+	
+	public Set<String> names() {
```
For terminology consistency with the code below, you should probably call this `getIdentifiers`

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/EnumValue.java:
```diff
@@ -0,0 +1,56 @@
+	private Value enumValue;
+	
+	public EnumValue(EnumTypeValue enumType, String enumIdentifier) throws NoSuchEnumIdentifierException {
+		this.enumType = enumType;
```
That this property is called `enumType` as well as the property on `EnumTypeValue` might cause some confusion. Not sure what the best name is... perhaps just `type`?

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/EnumValue.java:
```diff
@@ -0,0 +1,56 @@
+			return enumIdentifier.equals(((EnumValue) obj).enumIdentifier);
+		}
+		
+		return getValue().equals(obj);
```
I'm not sure this comparison is sufficient. Don't you also need to make sure they are of the same `EnumTypeValue`?

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/Expression.java:
```diff
@@ -3,7 +3,7 @@
 	public abstract TypeValue getType();
 	public abstract Value evaluate();
-	public abstract void verify();
+	public abstract void verify() throws Exception;
```
Oops! Good catch :+1:

lucaswoj commented:
This is **awesome** structurally. Just some pedantic naming feedback. :blush:  This is the most complicated type we've made yet (pending, of course, figuring out #24)!

maxqchen committed:
Renamed several values
Changed EnumValue's equals() to check for type as well for strict enum
comparison

maxqchen commented:
Okay, made a few renames and updated the equals method. 

I think equality at this point is still incomplete. We're going to need something like Values.equals(Value v1, Value v2) method.

lucaswoj commented:
@maxqchen Agreed. We're going to need some standard methods to check for equality, types, etc.

lucaswoj commented:
@maxqchen  Really keen to get this merged. One last tiny tweak!

maxqchen committed:
Merge branch 'intermediate' of https://github.com/MurphEngineer/whdl

maxqchen committed:
added guava as a dependency

maxqchen commented:
I don't think we need this anymore, right? Going to close it in that case.

lucaswoj commented:
I think this code is still useful to us. Lets merge it before anyone tries to do #83 :wink:

lucaswoj commented:
Can't see any way to re-open. Perhaps your fork repo moved?

maxqchen commented:
Yeah, with all the repo changes I re-made my fork from master. Was under the impression that with our spec change we weren't going to use this code anymore.

lucaswoj commented:
Nothing in the front-end language is really changing much, except for some more generalized names. Can you try to salvage this pull request?

maxqchen commented:
Yeah my bad then, I'll redo this req.
